### ISSUE TITLE
Code of your "Learning Hard Retrieval Decoder Attention for Transformers" paper

### ISSUE THREAD

**Issue by HaisongDing** [hfxunlp/transformer#3]
Hi Hongfei, 
Does this repo also contain the implementation of your "Learning Hard Retrieval Decoder Attention for Transformers" paper? If not, will it be released? Based on my understanding, the "hard retrieval" is achieved by replacing P with P'=Multinomial Sampling(P), P=(P'-P).detach()+P. Please kindly correct me if I am wrong.

**Comment by hfxunlp** [comment]
@HaisongDing 
Hi Haisong,
many thanks for your attention. Please find the implementation of the attention modules [here](https://github.com/hfxunlp/transformer/blob/master/modules/attn/retr.py), and the retrieval function [here](https://github.com/hfxunlp/transformer/blob/master/modules/sampler.py). These python code exactly implements the equations in the paper for efficiency, but yours looks simpler and more clever.
Besides, it is suggested to implement the training and inference (argmax+indexing) differently, as this benefits the inference performance.
Best,
Hongfei

**Comment by HaisongDing** [comment]
Thanks for your quick response.
